gcloud-java-storage/src/main/java/com/google/gcloud/storage/testing/FakeStorageRpc.java

```diff
@@ -0,0 +1,264 @@
+        // this is a bit of a hack, since we don't implement generations.
+        if ((StorageRpc.Option) option == Option.IF_GENERATION_MATCH
+            && ((Long) options.get(option)).longValue() == 0L) {
+          mustNotExist = true;
+        }
+      }
+    }
+    if (mustNotExist && stuff.containsKey(key)) {
+      throw new StorageException(new java.nio.file.FileAlreadyExistsException(key));
+    }
+    stuff.put(key, object);
+
+    return fullname(object);
+  }
+
+  @Override
+  public void write(String uploadId, byte[] toWrite, int toWriteOffset, long destOffset,
+                    int length, boolean last) throws StorageException {
+    // this may have a lot more allocations than ideal, but it'll work.
+    byte[] bytes;
+    if (contents.containsKey(uploadId)) {
+      bytes = contents.get(uploadId);
+      if (bytes.length < length + destOffset) {
+        bytes = new byte[(int) (length + destOffset)];
+      }
+    } else {
+      bytes = new byte[(int) (length + destOffset)];
+    }
+    System.arraycopy(toWrite, toWriteOffset, bytes, (int) destOffset, length);
+    contents.put(uploadId, bytes);
+  }
+
+  @Override
+  public RewriteResponse openRewrite(RewriteRequest rewriteRequest) throws StorageException {
+    String sourceKey = fullname(rewriteRequest.source);
+    String destKey = fullname(rewriteRequest.target);
+    boolean mustNotExist = false;
+    // a little hackish, just good enough for the tests to work.
+    if (!contents.containsKey(sourceKey)) {
+      throw new StorageException(new NoSuchFileException(sourceKey));
+    }
+
+    for (Option option : rewriteRequest.targetOptions.keySet()) {
+      if (option instanceof StorageRpc.Option) {
+        // this is a bit of a hack, since we don't implement generations.
+        if ((StorageRpc.Option) option == Option.IF_GENERATION_MATCH
+            && ((Long) rewriteRequest.targetOptions.get(option)).longValue() == 0L) {
+          mustNotExist = true;
+        }
+      }
+    }
+
+    if (mustNotExist && contents.containsKey(destKey)) {
+      throw new StorageException(new FileAlreadyExistsException(destKey));
+    }
+
+    stuff.put(destKey, rewriteRequest.target);
+
+    byte[] data = contents.get(sourceKey);
+    contents.put(destKey, Arrays.copyOf(data, data.length));
+    return new RewriteResponse(rewriteRequest, rewriteRequest.target, data.length, true,
+        "rewriteToken goes here", data.length);
+  }
+
+  @Override
+  public RewriteResponse continueRewrite(RewriteResponse previousResponse) throws StorageException {
+    throw new NotImplementedException();
+  }
+
+  // setFullName sets the bucket and name.
+  private void setFullName(StorageObject toModify, String fullname) {
+    int sep = fullname.indexOf('/', 1);
+    toModify.setBucket(fullname.substring(0, sep));
+    toModify.setName(fullname.substring(sep + 1));
+  }
+
+  private String fullname(StorageObject so) {
+    return (so.getBucket() + "/" + so.getName()).replaceAll("//", "/");
```

Comment:
Why do you do the replaceAll here? I would expect such cleanup to be done by the nio contrib part.